vmcheck/vmcheck_linux.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vmcheck
+
+import (
+	"encoding/binary"
+	"syscall"
+
+	"github.com/vmware/vmw-guestinfo/bdoor"
+)
+
+// From https://github.com/intel-go/cpuid/blob/master/cpuidlow_amd64.s
+// Get the CPU ID low level leaf values.
+func cpuid_low(arg1, arg2 uint32) (eax, ebx, ecx, edx uint32)
+
+// IsVirtualWorld returns true if running in a VM.
+func IsVirtualWorld() (bool, error) {
+	// Test the HV bit is set
+	if !CPUisVM() {
+		return false, nil
+	}
+
+	// Test if backdoor port is available.
+	if isVM, err := HypervisorPortCheck(); err != nil || !isVM {
+		return isVM, err
+	}
+
+	return true, nil
+}
+
+func HypervisorPortCheck() (bool, error) {
+	if err := syscall.Ioperm(int(bdoor.BackdoorPort), int(bdoor.BackdoorPort)+1, 1); err != nil {
+		return false, err
+	}
+
+	p := &bdoor.BackdoorProto{}
+	p.CX.Low.SetWord(bdoor.CommandGetVersion)
+
+	out := p.InOut()
+
+	// if there is no device, we get back all 1s
```

Comment:
I added this to my vmx and ensured the openvm-toolsd couldn't run.  When I'd start it, it would exit immediately and quietly.  However,  the port is still available and we're able to talk to `hostd` through it. 

```
$ TEST_TOOLBOX=1 sudo -E go test -v ./...
=== RUN   TestBdoorArgAlignment
--- PASS: TestBdoorArgAlignment (0.00s)
=== RUN   TestSetWord
--- PASS: TestSetWord (0.00s)
=== RUN   TestQuadToHighLow
--- PASS: TestQuadToHighLow (0.00s)
=== RUN   TestHighLowToQuad
--- PASS: TestHighLowToQuad (0.00s)
PASS
ok  	github.com/vmware/vmw-guestinfo/bdoor	0.001s
?   	github.com/vmware/vmw-guestinfo/examples	[no test files]
=== RUN   TestOpenClose
2017/01/23 00:02:58 Opened channel 0
2017/01/23 00:02:58 sending "info" over 0
2017/01/23 00:02:58 sending "-get" over 0
2017/01/23 00:02:58 sending " gue" over 0
2017/01/23 00:02:58 sending "stin" over 0
2017/01/23 00:02:58 sending "fo.d" over 0
2017/01/23 00:02:58 sending "oesn" over 0
2017/01/23 00:02:58 sending "otex" over 0
2017/01/23 00:02:58 sending "istd" over 0
2017/01/23 00:02:58 sending "oesn" over 0
2017/01/23 00:02:58 sending "otex" over 0
2017/01/23 00:02:58 sending "it" over 0
2017/01/23 00:02:58 Received 0x6f4e2030
2017/01/23 00:02:58 Received 0x6c617620
2017/01/23 00:02:58 Received 0x66206575
2017/01/23 00:02:58 Received 0x646e756f
2017/01/23 00:02:58 Closed channel 0
2017/01/23 00:02:58 Opened channel 0
2017/01/23 00:02:58 Closed channel 0
--- PASS: TestOpenClose (0.00s)
=== RUN   TestReset
2017/01/23 00:02:58 Opened channel 0
2017/01/23 00:02:58 Closed channel 0
--- PASS: TestReset (0.00s)
PASS
ok  	github.com/vmware/vmw-guestinfo/message	0.007s
?   	github.com/vmware/vmw-guestinfo/rpcout	[no test files]
?   	github.com/vmware/vmw-guestinfo/rpcvmx	[no test files]
?   	github.com/vmware/vmw-guestinfo/util	[no test files]
=== RUN   TestIsVirtualWorld
--- PASS: TestIsVirtualWorld (0.00s)
	vmcheck_test.go:29: Backdoor available:  true
	vmcheck_test.go:30: CPU HV:  true
	vmcheck_test.go:35: Running in a VM:  true
PASS
ok  	github.com/vmware/vmw-guestinfo/vmcheck	0.006s
```